### ISSUE TITLE
Support for other repository management tools

### ISSUE THREAD

**Issue by stil4m** [gdotdesign/elm-github-install#4]
Are there ideas/plans to support other repository management tools such as:

* Bitbucket
* Bitbucket Server
* GitLab
* ...

I am willing to implement this, but how would this reflect on the current functionality?

**Comment by gdotdesign** [comment]
Hey, thanks for bringing this up.

There are a few things that needs to be resolved if we want to support other git based services:
* Compatibility with `elm-pakage`: right now `elm-github-install` can be used as a direct replacement on packages that doesn't use Github packages and I'd like to keep it that way (defaulting to github)
* Distinguish between services
* How packages are downloaded ( right now it uses direct downloading ) 
* How `elm-package.json` is downloaded for a given version ( right now it uses direct downloading ) 
* Don't degrade performance much

I don't know If these can be done with just **Git** (I just have a sallow understanding of it) but if it can that would be ideal.

If we are touching this subject we can think about handling other references (not just tags) but `semver-resolver` seems to be limited to semver only.

**Comment by vankeisb** [comment]
Hello ! 

Cool tool you've written, thank you for sharing. 

I also needed to use alternate "repositories", for obvious corporate reasons. I have hacked your script a little bit to make this work...

It uses an (optional) "semicolon" syntax in order to use git+ssh instead of the actual https connection, and provide a different host name :

```
# elm-package.json
{
  ...
  "dependencies": {
    ...
    "github.internal.myco.com:NoRedInk/nri-elm-css": "1.3.0 <= 1.3.0 < 2.0.0",
    ...
  }
  ...
}
```

In this example, the dependency is fetched from the repo :

    git@github.internal.myco.com:NoRedInk/nri-elm-css.git 

It uses git+ssh, so that it works with ssh keys, without https at all (only git commands). We have github entreprise, but it is behind a SSO, and unauthenticated http download doesn't work. 

The git+ssh version is much slower than the default https one, but it's acceptable for us. The script is not meant to be used in production but at build-time, so it's ok to wait a few secs more.

**Comment by vankeisb** [comment]
I'm preparing a PR...

**Comment by gdotdesign** [comment]
Hi folks, 

I've been working on this in the last two days, and I've come up with a new version.

I'm more comfortable with Ruby and it's ecosystem then NodeJS and it's ecosystem, and because of that I've recreated this from scratch in Ruby. 

This new version works a little differently but has a lot more features and functionality:
* It's no longer Github dependent:
  * it works with any Git repository
  * it supports any Git reference as a version (provided there is an `elm-package.json` at that state)
* It keeps local clones of the repositories (in ~/.elm-install) and because of that:
  * It can **install packages offline** provided they are already cloned
  * It's slow in the first few installs but really quick after that
  * The dependency graph is cached locally

I released a **very preliminary version** of it on RubyGems https://rubygems.org/gems/elm_install which can be installed with `gem install elm_install`and also a packaged runnable version at the release page https://github.com/gdotdesign/elm-github-install/releases/tag/v0.1.0

How to use it
==========
* Package names accept any valid git url (git://*, https://*, git@*) as package name.
* Non version references such as branches or commit hashes can be used as a version prefixed by **ref:**
* Normal package names such as `elm-lang/core` are used as `git@github.com/elm-lang/core`, so it falls back to use Github.

All of these are valid and point to the same code:
```
"dependencies": {
  "gdotdesign/elm-ui": "0.4.0",
  "git@github.com:gdotdesign/elm-ui.git": "ref:master",
  "https://github.com/gdotdesign/elm-ui.git": "ref:fca5a61f7867a9f975cffa27e669b08b846745db"
}
```

Also the name and use of command / binary is changed from `elm-github-install` to `elm-install install`.

I'm curious about what do you think about this in general (using Ruby, using the clones, renaming the command etc..), and I hope some of you will try it out and give some feedback.

**Comment by vankeisb** [comment]
Looks great ! I will definitely try it out. 
Could it handle snapshots (grab HEAD of a given branch) ? This way we would not need a tag ? Would be useful when depending on your own modules.

**Comment by gdotdesign** [comment]
Yes it can `ref:master`, `ref:development` or `ref:commit_hash` instead of the version works, but it needs an valid `elm-package.json` at that ref.

**Comment by stil4m** [comment]
Great work!

**Comment by stil4m** [comment]
I'm testing this out, but when I run elm-make I get errors such as:

```
Could not find package git@github.com:gdotdesign/elm-ui.git.

Maybe your elm-stuff/ directory has been corrupted? You can usually fix stuff
like this by deleting elm-stuff/ and rebuilding your project.
```

Would it be an idea to change the setup of the `elm-package.json` to a format like:

```
{
...
  "dependencies" : {
    "gdotdesign/elm-ui": "0.4.0 <= v < 1.0.0",
    "elm-lang/core" "5.0.0 <= v < 6.0.0"
  },
  "dependencyResolving" : {
    "gdotdesign/elm-ui" : "git@github.com:gdotdesign/elm-ui.git"
  }
...
}
```

This might make it less likely that the package format will conflict with the default elm tooling.

**Comment by gdotdesign** [comment]
Yes, you were right, I didn't test it properly.

I've managed to make it work at last :smile: I also tested it with a Bitbutcket repository along with using the **master branch** from **elm-lang/core**.

This version uses the `dependency-sources` field in the `elm-package.json`. It can either be an URL or a hash with `url` and `ref`.

I've tested it with this `elm-package.json`:
```
{
  "version": "1.0.0",
  "summary": "Test",
  "repository": "https://github.com/user/repo.git",
  "license": "BSD3",
  "native-modules": true,
  "source-directories": [
    "."
  ],
  "exposed-modules": [
  ],
  "dependencies": {
    "gdotdesign/elm-install-test": "1.0.0 <= v < 2.0.0",
    "elm-lang/core": "5.0.0 <= v < 6.0.0",
    "elm-lang/svg": "2.0.0 <= v < 3.0.0",
    "elm-lang/dom": "1.1.1 <= v < 2.0.0"
  },
  "dependency-sources": {
    "elm-lang/core": {
      "url": "git@github.com:elm-lang/core",
      "ref": "master"
    },
    "gdotdesign/elm-install-test": {
      "url": "https://gdotdesign@bitbucket.org/gdotdesign/elm-install-test",
      "ref": "master"
    }
  },
  "elm-version": "0.18.0 <= v < 0.19.0"
}
```

Running `elm-make` after installation works without errors.

I've released it as `0.1.2` (binaries on the releases page), let me know if is there any issues.

**Comment by stil4m** [comment]
Cool. Nice work. I'll give it a try tomorrow

**Comment by gdotdesign** [comment]
#7 This is the PR for this.

**Comment by stil4m** [comment]
I've just started testing. We're getting there, but I've found another bug in the script. In the company I work we host a private bitbucket. Two sample urls for a repo would be:

```
git@<ourdomain>:7999/~<my-username>/<my-repo>.git
git@<ourdomain>:7999/<project-name>/<the-repo>.git
```

With the newest version (0.1.3) I get the following error: 

```
  Resolving packages...

  The source of package foo/bar is set to git@<our-domain>:7999/~<my-username>/<my-repo>.git which would
  be install to elm-stuff/7999/~<my-username>/<my-repo>.git. This would cause a conflict
  when trying to compile anything.

  The name of a package must match the source url's path.

  foo/bar <=> 7999/~<my-username>/<my-repo>.git
```

My config is:

```
    "dependencies": {
        <other-packages>,
        "foo/bar": "1.0.0 <= v < 2.0.0"
    },
    "dependency-sources": {
      "foo/bar" : {
        "url": "git@<our-domain>:7999/~<my-username>/<my-repo>.git",
        "ref": "master"
      }
    },
```

I believe the script current version assumes everything after the colon will be the project name. Which is the case for GitHub and Bitbucket cloud, but may conflict with other Git hosts. For us this is a port number. Additionally we have a tilde for a username. 

I understand why you would want to clone to the folder `$HOME/.elm-install/foo/bar`, because of the conflicts with other repositories. Can't you trust the user to choose a name for a repository that is not used?

Would be storing the clones in `$CWD/.elm-install` be a more cleaner solution? This would mean more that git clones will be applied, but it would be more nicer for CI, because you can easier clean/clear the directory. Additionally this would lead to less conflicting repositories (less chance within one project).

I would really like to help you with this, but my Ruby skills are not that good, and I do not know which direction you want to take this. But let me know if I can help via this or Slack.

**Comment by gdotdesign** [comment]
Thanks for trying it out, it really helps that it's tested in a live environment!

1. The port is really a bug in the gem I choose to parse git URLs, it doesn't seem to recognize them, so I'll probably have to write that from scratch too (it's not a very complicated thing). 

2. The tilde doesn't effect the default Elm tooling: I updated the directories and files by hand after an installation and everything seems to work, you just have to add it everywhere.

3. The cache directory can be specified in the code, it's not exposed right now but it will be, so you could put it where you wished.

I've tried to set up a local git server but it's one of the things I'm not experienced at, so I really cannot test these edge cases, so the best way to help is to just to try new versions and reporting problems as they come up :smile:

I'll let you know once I have a fix for the port.

**Comment by gdotdesign** [comment]
Do you happen to know a public git host which I can test the ports with? 

I've tried it with Github but didn't seem to make it work:
```
git clone git@github.com:22/elm-lang/core.git test

Cloning into 'test'...
fatal: remote error: 
  22/elm-lang/core is not a valid repository name
  Email support@github.com for help
```

**Comment by stil4m** [comment]
I do not know one from the top of my head, besides my companies private bitbucket :(. In addition I know that code base uses a three part path:

```
git@codebasehq.com:<organisation>/<project>/<repo-name>.git
```

**Comment by gdotdesign** [comment]
OK so I did a little digging about which protocols / URLs can be used in git: https://www.kernel.org/pub/software/scm/git/docs/git-clone.html (GIT URLS section). 

As it turns out the SCP url does not support ports so this is invalid `git@github.com:22/elm-lang/core.git` but SCP is just a shorthand for the **ssh://** URL, which is supported in v0.1.3.

I've tried it with `ssh://git@github.com:22/gdotdesign/elm-dom` where **22** is the SSH port.

So in your case the files would look something like this:
```
"dependencies": {
  <other-packages>,
  "~foo/bar": "1.0.0 <= v < 2.0.0"
},
"dependency-sources": {
  "foo/bar" : {
    "url": "ssh://git@<our-domain>:7999/~foo/bar",
    "ref": "master"
  }
},
```

**Comment by stil4m** [comment]
That makes sense. I got it to work 👍 . One addition, which may be something for a FAQ: My ssh url was suffixed with `.git`. I got that directly from the Git host. I had to remove this to make it work (`<x>.git ==> <x>`) .

**Comment by stil4m** [comment]
I've found one more thing. May be something for a FAQ too, or a change in the implementation. If you have no SSH registered at GitHub, you cant clone repositories via `git@github.com:<x>/<y>`. Instead, you are able to clone `https://github.com/<x>/<y>.git`. Would it make more sense to use the `HTTPS` protocol for the default packages? Otherwise, I think this should be documented properly (I can do this if you want).

**Comment by gdotdesign** [comment]
Thanks, I'm glad it's working now.

Both your points are valid, I'll change the default to use `https` not everyhone has SSH access.

**Comment by gdotdesign** [comment]
I'll released the ruby version as 0.3.0: https://github.com/gdotdesign/elm-github-install/releases/tag/v0.3.0

* The npm package is still there, I've managed to install the standalone versions though npm install scripts
* `.git` ending is now supported
* Changed the Github default to `https`
